Doc/library/contextvars.rst

```diff
@@ -144,21 +144,89 @@ Manual Context Management
+   *Entering* a Context, either by calling the
+   :meth:`Context.run` method or using the Context object as a :term:`context
+   manager`, pushes the Context onto the top of the current thread's stack,
+   making it the current Context.  *Exiting* from the current Context, either by
+   returning from the callback passed to :meth:`Context.run` or by exiting the
+   :keyword:`with` statement suite, pops the Context off of the top of the
+   stack, restoring the current Context to what it was before.
+
+   Because each thread has its own Context stack, :class:`ContextVar` objects
+   behave in a similar fashion to :func:`threading.local` when values are
+   assigned in different threads.
+
+   Attempting to do either of the following will raise a :exc:`RuntimeError`:
+
+   * Entering an already entered Context.  (This includes Contexts entered in
+     other threads.)
+   * Exiting from a Context that is not the current Context.
+
+   After exiting a Context, it can later be re-entered (from any thread).
+
+   Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
+   method are recorded in the current Context.  The :meth:`ContextVar.get`
+   method returns the value associated with the current Context.  Thus, exiting
+   a Context effectively reverts any changes made to context variables while the
+   Context was entered.  (If desired, the values can be restored by re-entering
+   the Context.)
 
    Context implements the :class:`collections.abc.Mapping` interface.
 
+   .. versionadded:: 3.14
+      Context implements the :term:`!context management protocol` (:pep:`343`).
+
+   .. method:: __enter__(self)
+               __exit__(self, exc_type, exc_value, exc_tb)
+
+      These methods implement the :term:`!context management protocol`
+      (:pep:`343`), making it possible to use a Context object as a
+      :term:`context manager` with the :keyword:`with` statement.  The
+      :meth:`__enter__` method enters the Context and returns *self*, so the
+      value bound to the identifier given in the :keyword:`with` statement's
+      :keyword:`!as` clause (if present) is the Context object itself.  The
+      :meth:`__exit__` method exits the Context.
+
+      Example:
+
+      .. testcode::
+
+         import contextvars
+
+         var = contextvars.ContextVar("var")
+         var.set("initial")
+         assert var.get() == "initial"
+
+         # Copy the current Context and enter the copy.
+         with contextvars.copy_context() as ctx:
+             var.set("updated")
+             assert var in ctx
+             assert ctx[var] == "updated"
+             assert var.get() == "updated"
+
+         # Exited ctx, so the observed value of var has reverted.
+         assert var.get() == "initial"
+         # But the updated value is still recorded in ctx.
+         assert ctx[var] == "updated"
+
+         # Re-entering ctx restores the updated value of var.
+         with ctx:
+             assert var.get() == "updated"
```

Comment:
Note the use of `.. testcode::`.  I want this example to be testable to prevent the implementation from drifting away from the documentation.  Would you prefer I use `print` and `.. testoutput::`?  I personally find `assert` easier to read.

Reply:
Is that test code actually checked, though? I guess put differently: can you point me at the CPython CI that's actually executing this test code and checking it is as expected?

Reply:
I updated the `run` documentation to also use `testcode`/`testoutput`.